extra/modules/greenbids-real-time-data/src/main/java/org/prebid/server/hooks/modules/greenbids/real/time/data/config/DatabaseReaderFactory.java

```diff
@@ -1,52 +1,129 @@
 package org.prebid.server.hooks.modules.greenbids.real.time.data.config;
 
+import io.netty.handler.codec.http.HttpResponseStatus;
+import io.vertx.core.http.HttpClientRequest;
+import io.vertx.core.http.HttpClientResponse;
+import com.maxmind.db.Reader;
+import io.vertx.core.Future;
 import io.vertx.core.Promise;
 import io.vertx.core.Vertx;
 import com.maxmind.geoip2.DatabaseReader;
+import io.vertx.core.file.OpenOptions;
+import io.vertx.core.http.HttpMethod;
+import io.vertx.core.http.RequestOptions;
+import org.apache.commons.compress.archivers.tar.TarArchiveEntry;
+import org.apache.commons.compress.archivers.tar.TarArchiveInputStream;
 import org.prebid.server.exception.PreBidException;
+import org.prebid.server.log.Logger;
+import org.prebid.server.log.LoggerFactory;
 import org.prebid.server.vertx.Initializable;
 
-import java.io.FileOutputStream;
 import java.io.IOException;
-import java.io.InputStream;
-import java.net.URL;
 import java.nio.file.Files;
 import java.nio.file.Path;
 import java.util.concurrent.atomic.AtomicReference;
+import java.util.zip.GZIPInputStream;
 
 public class DatabaseReaderFactory implements Initializable {
 
-    private final String geoLiteCountryUrl;
+    private final GreenbidsRealTimeDataProperties properties;
 
     private final Vertx vertx;
 
     private final AtomicReference<DatabaseReader> databaseReaderRef = new AtomicReference<>();
 
-    public DatabaseReaderFactory(String geoLitCountryUrl, Vertx vertx) {
-        this.geoLiteCountryUrl = geoLitCountryUrl;
+    private static final Logger logger = LoggerFactory.getLogger(DatabaseReaderFactory.class);
+
+    public DatabaseReaderFactory(
+            GreenbidsRealTimeDataProperties properties, Vertx vertx) {
+        this.properties = properties;
         this.vertx = vertx;
     }
 
     @Override
     public void initialize(Promise<Void> initializePromise) {
+        vertx.executeBlocking(() -> downloadAndExtract().onSuccess(databaseReaderRef::set))
+                .<Void>mapEmpty()
+                .onComplete(initializePromise);
+    }
+
+    private Future<DatabaseReader> downloadAndExtract() {
+        final String downloadUrl = properties.geoLiteCountryPath;
+        final String tmpPath = properties.tmpPath;
+        return downloadFile(downloadUrl, tmpPath)
+                .compose(unused -> {
```

Comment:
just map?

Reply:
Replaced .compose with .flatMap to convert `Future<Future<DatabaseReader>>` into `Future<DatabaseReader>`